crypto/s2n_libcrypto.c

```diff
@@ -13,14 +13,73 @@
  * permissions and limitations under the License.
  */
 
+#include <openssl/crypto.h>
+
 #include "crypto/s2n_crypto.h"
 #include "crypto/s2n_openssl.h"
+#include "crypto/s2n_libcrypto.h"
+#include "utils/s2n_safety_macros.h"
 
-/*
- * Verifying libcrypto type via method should be preferred
- * where possible, since it reduces #ifs and avoids potential
- * bugs where the header containing the #define is not included.
+#include <string.h>
+
+#define EXPECTED_AWSLC_VERSION_NAME "AWS-LC"
+#define EXPECTED_BORINGSSL_VERSION_NAME "BoringSSL"
+
+/* https://www.openssl.org/docs/man{1.0.2, 1.1.1, 3.0}/man3/OPENSSL_VERSION_NUMBER.html
+ * OPENSSL_VERSION_NUMBER in hex is: MNNFFPPS major minor fix patch status.
+ * Bitwise: MMMMNNNNNNNNFFFFFFFFPPPPPPPPSSSS
+ * To not be overly restrictive, we onyl care about major and minor versions.
+ * From OpenSSL 3.0 the "fix" part is also deprecated and is always a flat 0x00.
  */
+#define VERSION_NUMBER_MASK 0xFFF00000L
+
+/* Returns the version name of the libcrypto containing the definition that the
+ * symbol OpenSSL_version binded to at link-time. This can be used as
+ * verification at run-time that s2n linked against the expected libcrypto.
+ */
+const char * s2n_libcrypto_get_version_name(void)
+{
+    return OpenSSL_version(OPENSSL_VERSION);
+}
+
+static S2N_RESULT s2n_libcrypto_validate_expected_version_name(const char *expected_version_name)
+{
+    RESULT_ENSURE_REF(expected_version_name);
+    RESULT_ENSURE_REF(s2n_libcrypto_get_version_name());
+    RESULT_ENSURE_EQ(strlen(expected_version_name), strlen(s2n_libcrypto_get_version_name()));
+    RESULT_ENSURE(memcmp(expected_version_name, s2n_libcrypto_get_version_name(), strlen(expected_version_name)) == 0, S2N_ERR_LIBCRYPTO_VERSION_NAME_MISMATCH);
```

Comment:
linter is flagging `memcmp` as well

Reply:
Using `s2n_constant_time_equals()` now, even though it requires some casting...